src/org/rascalmpl/util/maven/Artifact.java

```diff
@@ -130,70 +133,159 @@ private static boolean isVersionRange(String version) {
+            } else {
+                alreadyResolved = false;
+                resolvedVersions.put(versionLess, version);
+            }
+
+            var key = new ResolveKey(coordinate, excludes);
+            if (alreadySolved.contains(key) || !d.shouldInclude(forScope)) {
                 continue;
             }
+            alreadySolved.add(key);
+
             if (d.getScope() == Scope.PROVIDED) {
                 // current maven behavior seems to be:
                 // - do not download provided dependencies
                 // - if a provided dependency is present in the maven repository it's considered "provided"
                 var pomDep = ourResolver.calculatePomPath(d.getCoordinate());
                 if (!Files.notExists(pomDep)) {
                     // ok, doesn't exist yet. so don't download it to calculate dependencies
-                    // and don't add it to the list since somewhere else somebody might depend on it 
+                    // and don't add it to the list since somewhere else somebody might depend on it
                     continue;
                 }
             }
             if (d.getScope() == Scope.SYSTEM) {
-                result.add(createSystemArtifact(d));                
+                if (!alreadyResolved) {
+                    result.add(createSystemArtifact(d));
+                }
                 continue;
             }
 
-            var art = parser.parseArtifact(coordinate, d.getExclusions(), ourResolver);
+            Set<WithoutVersion> newExcludes = excludes;
+            if (!d.getExclusions().isEmpty()) {
+                newExcludes = new HashSet<>(excludes);
+                newExcludes.addAll(d.getExclusions());
+            }
+
+            var art = parser.parseArtifact(coordinate, newExcludes, ourResolver);
             if (art != null) {
-                result.add(art);
-                nextLevel.add(art);
+                if (!alreadyResolved) {
+                    result.add(art);
+                }
+                nextLevel.add(new ResolveState(art, newExcludes));
             }
-            alreadyIncluded.add(withoutVersion);
         }
         if (forScope == Scope.TEST) {
             // only do test scope for top level, switch to compile after that
             forScope = Scope.COMPILE;
         }
-        // now we go through the new artifacts and make sure we add their dependencies if needed 
+        // now we go through the new artifacts and make sure we add their dependencies if needed
         // but now in their context (their resolver etc)
-        for (var a: nextLevel) {
-            a.calculateClassPath(forScope, alreadyIncluded, rangedDeps, result, parser);
+        for (var state: nextLevel) {
+            state.calculateClassPath(forScope, alreadySolved, resolvedVersions, rangedDeps, result, parser);
```

Comment:
do we really need this state? isn't the recursion enough to solve this? as in, our own stack is already the state we need? I'm not sure about this comment, but I'm just checking to make sure it's not more heavy than needed.